gensim/models/poincare.py

```diff
@@ -731,27 +739,457 @@ def compute_loss(self):
-        norm_2 = np.linalg.norm(vector_2)
-        euclidean_dist = np.linalg.norm(vector_1 - vector_2)
-        if euclidean_dist == 0.0:
-            return 0.0
-        else:
-            return np.arccosh(
-                1 + 2 * (
-                    (euclidean_dist ** 2) / ((1 - norm_1 ** 2) * (1 - norm_2 ** 2))
-                )
+    def poincare_distance(vector_1, vector_2):
+        """
+        Return poincare distance between two input vectors. Convenience method over `poincare_distance_batch`.
+
+        Parameters
+        ----------
+        vector_1 : numpy.array
+            input vector
+        vector_2 : numpy.array
+            input vector
+
+        Returns
+        -------
+        numpy.float
+            Contains Poincare distance between `vector_1` and `vector_2`.
+
+        """
+        return PoincareKeyedVectors.poincare_distance_batch(vector_1, vector_2[np.newaxis, :])[0]
+
+    @staticmethod
+    def poincare_distance_batch(vector_1, vectors_all):
+        """
+        Return poincare distances between one vector and a set of other vectors.
+
+        Parameters
+        ----------
+        vector_1 : numpy.array
+            vector from which Poincare distances are to be computed.
+            expected shape (dim,)
+        vectors_all : numpy.array
+            for each row in vectors_all, distance from vector_1 is computed.
+            expected shape (num_vectors, dim)
+
+        Returns
+        -------
+        numpy.array
+            Contains Poincare distance between vector_1 and each row in vectors_all.
+            shape (num_vectors,)
+
+        """
+        euclidean_dists = np.linalg.norm(vector_1 - vectors_all, axis=1)
+        norm = np.linalg.norm(vector_1)
+        all_norms = np.linalg.norm(vectors_all, axis=1)
+        return np.arccosh(
+            1 + 2 * (
+                (euclidean_dists ** 2) / ((1 - norm ** 2) * (1 - all_norms ** 2))
             )
-    # TODO: Add other KeyedVector supported methods - most_similar, etc.
+        )
+
+    def nodes_closer_than(self, term_1, term_2):
+        """
+        Returns all nodes that are closer to `term_1` than `term_2` from `term_1`.
+
+        Parameters
+        ----------
+        term_1 : str
+            Input term.
+        term_2 : str
+            Input term.
+
+        Returns
+        -------
+        List (str)
+            List of nodes that are closer to `term_1` than `term_2` is from `term_1`.
+
+        Examples
+        --------
+
+        >>> model.closer_than('carnivore.n.01', 'mammal.n.01')
+        ['dog.n.01', 'canine.n.02']
+
+        """
+        all_distances = self.distances(term_1)
+        term_1_index = self.vocab[term_1].index
+        term_2_index = self.vocab[term_2].index
+        closer_node_indices = np.where(all_distances < all_distances[term_2_index])[0]
+        return [self.index2word[index] for index in closer_node_indices if index != term_1_index]
+
+    def rank(self, term_1, term_2):
```

Comment:
This looks generic. Would it make sense to move to the parent class?

Reply:
Good idea, done